### PR TITLE
Add TableScan, ScanFilterProject memory tracking

### PR DESCRIPTION
**Pull request by haozhun**
Also added system memory tracking to PageSource. Tracking for RecordCursor has not been added.

### CODE REVIEW

**Comment by haozhun** [comment]
Updated

File: presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcPageSource.java [hunk 1]
```diff
@@ -83,13 +84,16 @@
     private int batchId;
     private boolean closed;
 
+    private final AggregatedMemoryContext systemMemoryUsage;
```
Comment (cberner) [inline presto-hive/src/main/java/com/facebook/presto/hive/orc/OrcPageSource.java]: Name this `systemMemoryContext` or `memoryContext`?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 2]
```diff
@@ -0,0 +1,524 @@
...
+
+    private File tempFile;
+    private TestPreparer testPreparer;
+
+    @BeforeClass
+    public void setUp()
+            throws Exception
+    {
+        tempFile = File.createTempFile("presto_test_topsfmt", "orc");
+        tempFile.delete();
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Why do you need to delete it first?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 3]
```diff
@@ -0,0 +1,524 @@
...
+    private final int numRows = 50000;
+
+    private File tempFile;
+    private TestPreparer testPreparer;
+
+    @BeforeClass
+    public void setUp()
+            throws Exception
+    {
+        tempFile = File.createTempFile("presto_test_topsfmt", "orc");
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: What's "topsfmt"?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 4]
```diff
@@ -0,0 +1,524 @@
...
+
+    @Test
+    public void testPageSource()
+            throws Exception
+    {
+        ConnectorPageSource pageSource = testPreparer.newPageSource();
+
+        assertEquals(pageSource.getSystemMemoryUsage(), 0);
+
+        //*/
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Remove this

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 5]
```diff
@@ -0,0 +1,524 @@
...
+                VARCHAR.getSlice(block, block.getPositionCount() - 1); // trigger loading for lazy block
+                memoryUsage = pageSource.getSystemMemoryUsage();
+            }
+            else {
+                assertEquals(pageSource.getSystemMemoryUsage(), memoryUsage);
+                VARCHAR.getSlice(block, block.getPositionCount() - 1); // trigger loading for lazy block
+                assertEquals(pageSource.getSystemMemoryUsage(), memoryUsage);
+            }
+        }
+        assertBetweenInclusive(memoryUsage, 460000L, 469999L); // Memory usage after lazy-loading the actual block
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: Move this up to where you assign `memoryUsage`. It's hard to follow what this assertion is doing, and you only modify it in one place

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 6]
```diff
@@ -0,0 +1,524 @@
...
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: same comment

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 7]
```diff
@@ -0,0 +1,524 @@
...
+                VARCHAR.getSlice(block, block.getPositionCount() - 1); // trigger loading for lazy block
+                memoryUsage = pageSource.getSystemMemoryUsage();
+            }
+            else {
+                assertEquals(pageSource.getSystemMemoryUsage(), memoryUsage);
+                VARCHAR.getSlice(block, block.getPositionCount() - 1); // trigger loading for lazy block
+                assertEquals(pageSource.getSystemMemoryUsage(), memoryUsage);
+            }
+        }
+        assertBetweenInclusive(memoryUsage, 360000L, 369999L); // Memory usage after lazy-loading the actual block
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: same

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 8]
```diff
@@ -0,0 +1,524 @@
...
+            assertNotNull(page);
+            page.getBlock(1);
+            if (memoryUsage == -1) {
+                memoryUsage = driverContext.getSystemMemoryUsage();
+            }
+            else {
+                assertEquals(driverContext.getSystemMemoryUsage(), memoryUsage);
+            }
+        }
+        assertBetweenInclusive(memoryUsage, 460000L, 469999L);
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: same

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 9]
```diff
@@ -0,0 +1,524 @@
...
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: same

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 10]
```diff
@@ -0,0 +1,524 @@
...
+            assertNotNull(page);
+            page.getBlock(1);
+            if (memoryUsage == -1) {
+                memoryUsage = driverContext.getSystemMemoryUsage();
+            }
+            else {
+                assertEquals(driverContext.getSystemMemoryUsage(), memoryUsage);
+            }
+        }
+        assertBetweenInclusive(memoryUsage, 360000L, 369999L);
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: same

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 11]
```diff
@@ -0,0 +1,524 @@
...
+        for (int i = 52; i < 65; i++) {
+            assertFalse(operator.isFinished());
+            operator.getOutput();
+            assertBetweenInclusive(driverContext.getSystemMemoryUsage(), 500000L, 539999L);
+        }
+
+        // Page source is over, but data still exist in buffer of ScanFilterProjectOperator
+        Assert.assertFalse(operator.isFinished());
+        assertNull(operator.getOutput());
+        assertBetweenInclusive(driverContext.getSystemMemoryUsage(), 170000L, 179999L);
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: All these constants that are in here. Are they guaranteed by the file format, or might they change if we improve our reader in the future? If the latter, there should probably be a comment saying it's safe to change them if they break in the future

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 12]
```diff
@@ -0,0 +1,524 @@
...
+        }
+
+        for (int i = 52; i < 65; i++) {
+            assertFalse(operator.isFinished());
+            operator.getOutput();
+            assertBetweenInclusive(driverContext.getSystemMemoryUsage(), 500000L, 539999L);
+        }
+
+        // Page source is over, but data still exist in buffer of ScanFilterProjectOperator
+        Assert.assertFalse(operator.isFinished());
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: static import. Here and a lot of other places too

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 13]
```diff
@@ -0,0 +1,524 @@
...
+    }
+
+    private class TestPreparer
+    {
+        private final FileSplit fileSplit;
+        private final Properties schema;
+        private final List<HiveColumnHandle> columns;
+        private final List<Type> types;
+        private final List<HivePartitionKey> partitionKeys;
+        private final ExecutorService executor = Executors.newCachedThreadPool(daemonThreadsNamed("test-%s"));
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: static import

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 14]
```diff
@@ -0,0 +1,524 @@
...
+                            .collect(Collectors.joining(",")));
+            schema.setProperty(FILE_INPUT_FORMAT, OrcInputFormat.class.getName());
+            schema.setProperty(SERIALIZATION_LIB, serde.getClass().getName());
+
+            partitionKeys = testColumns.stream()
+                    .filter(TestColumn::isPartitionKey)
+                    .map(input -> new HivePartitionKey(input.getName(), HiveType.getHiveType(input.getObjectInspector()), (String) input.getWriteValue()))
+                    .collect(toList());
+
+            ImmutableList.Builder<HiveColumnHandle> columnsBuilder = new ImmutableList.Builder<>();
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: use the factory method for `Builder`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 15]
```diff
@@ -0,0 +1,524 @@
...
+            ImmutableList.Builder<Type> typesBuilder = new ImmutableList.Builder<>();
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: use the factory method for `Builder`

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 16]
```diff
@@ -0,0 +1,524 @@
...
+
+        public SourceOperator newTableScanOperator(DriverContext driverContext)
+        {
+            ConnectorPageSource pageSource = newPageSource();
+            SourceOperatorFactory sourceOperatorFactory = new TableScanOperatorFactory(
+                    0,
+                    new PlanNodeId("0"),
+                    (session, split, columnHandles) -> pageSource,
+                    types,
+                    columns.stream().map(columnHandle -> (ColumnHandle) columnHandle).collect(Collectors.toList())
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: static import

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 17]
```diff
@@ -0,0 +1,524 @@
...
+            );
+            SourceOperator operator = sourceOperatorFactory.createOperator(driverContext);
+            operator.addSplit(new Split("test", TestingSplit.createLocalSplit()));
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: static import?

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 18]
```diff
@@ -0,0 +1,524 @@
...
+                projectionsBuilder.add(singleColumn(types.get(i), i));
+            }
+            ImmutableList<ProjectionFunction> projections = projectionsBuilder.build();
+            SourceOperatorFactory sourceOperatorFactory = new ScanFilterAndProjectOperatorFactory(
+                    0,
+                    new PlanNodeId("0"),
+                    (session, split, columnHandles) -> pageSource,
+                    new GenericCursorProcessor(FilterFunctions.TRUE_FUNCTION, projections),
+                    new GenericPageProcessor(FilterFunctions.TRUE_FUNCTION, projections),
+                    columns.stream().map(columnHandle -> (ColumnHandle) columnHandle).collect(Collectors.toList()),
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: static import

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java [hunk 19]
```diff
@@ -0,0 +1,524 @@
...
+            );
+            SourceOperator operator = sourceOperatorFactory.createOperator(driverContext);
+            operator.addSplit(new Split("test", TestingSplit.createLocalSplit()));
+            return operator;
+        }
+
+        public SourceOperator newScanFilterAndProjectOperator(DriverContext driverContext)
+        {
+            ConnectorPageSource pageSource = newPageSource();
+            ImmutableList.Builder<ProjectionFunction> projectionsBuilder = new ImmutableList.Builder<>();
```
Comment (cberner) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: use factory

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryTracker.java [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.memory;
+
+public abstract class MemoryTracker
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryTracker.java]: What do you think of naming this `MemoryContext`?

File: presto-main/src/main/java/com/facebook/presto/memory/MemoryTracker.java [hunk 21]
```diff
@@ -0,0 +1,31 @@
...
+{
+    public abstract void reserveMemory(long bytes);
+
+    public abstract void freeMemory(long bytes);
+
+    public void updateSystemMemoryUsage(long deltaMemoryInBytes)
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/memory/MemoryTracker.java]: Consider removing "system" from this name

File: presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java [hunk 22]
```diff
@@ -173,7 +176,15 @@ public final boolean isFinished()
             finishing = true;
         }
 
-        return finishing && pageBuilder.isEmpty();
+        if (!finishing || !pageBuilder.isEmpty()) {
+            return false;
+        }
+
+        operatorContext.getSystemMemoryTracker().freeMemory(recordedPageBuilderMemoryUsage);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java]: I don't think this is the right place to be freeing system memory. `isFinished()` indicates that the operator is done, but this doesn't mean it's actually going to be garbage collected. I would probably put this in `Driver.destroyIfNecessary` where the other memory is released

File: presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java [hunk 23]
```diff
@@ -65,6 +65,9 @@
     private long completedBytes;
     private long readTimeNanos;
 
+    private long recordedPageSourceMemoryUsage;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/ScanFilterAndProjectOperator.java]: maybe name these "current" rather than "recorded"?

File: presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java [hunk 24]
```diff
@@ -197,6 +198,9 @@ public void finish()
                 throw Throwables.propagate(e);
             }
         }
+
+        operatorContext.getSystemMemoryTracker().freeMemory(recordedSystemMemoryUsage);
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java]: Don't assume the source freed all its memory. Get its memory usage and only free the delta, since it's not yet eligible for garbage collection (see my other comment about `isFinished`)

File: presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java [hunk 25]
```diff
@@ -108,6 +108,7 @@ public void close()
 
     private long completedBytes;
     private long readTimeNanos;
+    private long recordedSystemMemoryUsage;
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java]: "current" might be a better name than "recorded"

File: presto-orc/src/main/java/com/facebook/presto/orc/AggregatedMemoryContext.java [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+
+    public void close()
+    {
+        if (closed) {
+            return;
+        }
+        closed = true;
+        if (parentMemoryContext != null) {
+            parentMemoryContext.updateBytes(-usedBytes);
+        }
```
Comment (cberner) [inline presto-orc/src/main/java/com/facebook/presto/orc/AggregatedMemoryContext.java]: set `usedBytes = 0`

File: presto-orc/src/main/java/com/facebook/presto/orc/LocalMemoryContext.java [hunk 27]
```diff
@@ -0,0 +1,48 @@
...
+import static java.util.Objects.requireNonNull;
+
+public class LocalMemoryContext
+{
+    private final AggregatedMemoryContext parentMemoryContext;
+    private long usedBytes;
+
+    public LocalMemoryContext(AggregatedMemoryContext parentMemoryContext)
+    {
+        this.parentMemoryContext = requireNonNull(parentMemoryContext);
```
Comment (cberner) [inline presto-orc/src/main/java/com/facebook/presto/orc/LocalMemoryContext.java]: add error message string

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java [hunk 28]
```diff
@@ -122,13 +122,15 @@ public OrcReader(OrcDataSource orcDataSource, MetadataReader metadataReader)
 
         // read metadata
         Slice metadataSlice = completeFooterSlice.slice(0, metadataSize);
-        InputStream metadataInputStream = new OrcInputStream(orcDataSource.toString(), metadataSlice.getInput(), compressionKind, bufferSize);
-        this.metadata = metadataReader.readMetadata(metadataInputStream);
+        try (InputStream metadataInputStream = new OrcInputStream(orcDataSource.toString(), metadataSlice.getInput(), compressionKind, bufferSize, new AggregatedMemoryContext())) {
```
Comment (cberner) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcReader.java]: You have a lot of cleanups in this commit. I would move things like this to a separate commit

File: presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java [hunk 29]
```diff
@@ -62,13 +59,7 @@ public void testPageSource()
...
                 0,
                 new PlanNodeId("0"),
-                new PageSourceProvider() {
-                    @Override
-                    public ConnectorPageSource createPageSource(Session session, Split split, List<ColumnHandle> columns)
-                    {
-                        return new FixedPageSource(ImmutableList.of(input));
-                    }
-                },
+                (session, split, columns) -> new FixedPageSource(ImmutableList.of(input)),
```
Comment (cberner) [inline presto-main/src/test/java/com/facebook/presto/operator/TestScanFilterAndProjectOperator.java]: I'd separate these to a cleanup commit

File: presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java [hunk 30]
```diff
@@ -78,6 +86,7 @@ public void close()
             throws IOException
     {
         current = null;
+        systemMemoryUsage.setBytes(0);
```
Comment (cberner) [inline presto-orc/src/main/java/com/facebook/presto/orc/stream/OrcInputStream.java]: What about `buffer`? We didn't free that

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcPageSource.java [hunk 31]
```diff
@@ -111,6 +115,8 @@ else if (this.columnIndexes[i] == SHARD_UUID_COLUMN) {
                 constantBlocks[i] = buildSingleValueBlock(Slices.utf8Slice(shardUuid.toString()));
             }
         }
+
+        this.systemMemoryUsage = requireNonNull(systemMemoryUsage, "usedMemoryBytes is null");
```
Comment (cberner) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcPageSource.java]: wrong message

File: presto-spi/src/main/java/com/facebook/presto/spi/RecordCursor.java [hunk 32]
```diff
@@ -43,6 +43,11 @@
 
     boolean isNull(int field);
 
+    default long getSystemMemoryUsage()
+    {
+        return 16 * 1024 * 1024; // GUESSED_MEMORY_USAGE: 16MB
```
Comment (cberner) [inline presto-spi/src/main/java/com/facebook/presto/spi/RecordCursor.java]: I think it's fine to make this guess in the Hive connector, but I don't think the SPI should make a guess like this. I would just return zero

File: presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcPageSource.java [hunk 33]
```diff
@@ -82,7 +85,8 @@ public OrcPageSource(
             List<Long> columnIds,
             List<Type> columnTypes,
             List<Integer> columnIndexes,
-            UUID shardUuid)
+            UUID shardUuid,
+            AggregatedMemoryContext systemMemoryUsage)
```
Comment (cberner) [inline presto-raptor/src/main/java/com/facebook/presto/raptor/storage/OrcPageSource.java]: Instead of having this context which is shared with the `OrcRecordReader` why not just add a `getSystemMemoryUsage` to the reader and return that in `getSystemMemoryUsage`?

**Comment by cberner** [comment]
Sorry it took me so long to get to this!

File: presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java (hunk 2 above)
Comment (haozhun) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestOrcPageSourceMemoryTracking.java]: The file is created as soon as `File.createTempFile` is called. I copied from `OrcTester.TempFile`, which is a closeable wrapper for temp files. What I really need here is just a path. Using `createTempFile`+`delete` is a hack. Apparently, using temp file this way is bad (what if some one else decided to create the same file in the mean time). Therefore libraries won't support such use case.

**Comment by haozhun** [comment]
Addressed all comments except placement of asserts in `TestOrcPageSourceMemoryTracking`, which I forgot to talk to you in person.

**Comment by haozhun** [comment]
Addressed all comments.

File: presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java [hunk 34]
```diff
@@ -205,6 +209,9 @@ public boolean isFinished()
         if (!finished) {
             createSourceIfNecessary();
             finished = (source != null) && source.isFinished();
+            if (source != null) {
+                systemMemoryContext.setBytes(source.getSystemMemoryUsage());
```
Comment (cberner) [inline presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java]: Why do we need to update the memory usage here? It seems like the update in `finish()` will take care of it

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java [hunk 35]
```diff
@@ -365,9 +381,6 @@ private void advanceToNextStripe()
 
             rowGroups = stripe.getRowGroups().iterator();
         }
-        else {
-            rowGroups = ImmutableList.<RowGroup>of().iterator();
```
Comment (cberner) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Put this back? it looks like there's a code path that returns `null`

**Comment by cberner** [comment]
Couple minor comments, and refactor the contexts like we discussed. Otherwise looks good

**Comment by cberner** [comment]
Also, add comments to the temporary classes like `OperatorSystemMemoryContext` explaining that they'll be removed

File: presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java (hunk 34 above)
Comment (haozhun) [inline presto-main/src/main/java/com/facebook/presto/operator/TableScanOperator.java]: This need to exist because `source` may have been created 2 lines up. Also, `isFinished` could potentially change system memory usage.

File: presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java (hunk 35 above)
Comment (haozhun) [inline presto-orc/src/main/java/com/facebook/presto/orc/OrcRecordReader.java]: Addressed this comment.
